Port crm_claim_rma

aristobulo-eezee opened the pull request.

mistotebe committed:
[IMP] Port tests to v8 as well

Conflicts:
	crm_claim_rma/tests/test_lp_1282584.py

javier-eezee committed:
Update copyright. Update module descriptor to include Eezee-It

javier-eezee committed:
Updated crm_claim_rma to Odoo v8

aristobulo-eezee committed:
Move crm_claim_rma from __unported__

pedrobaeza commented on crm_claim_rma/account_invoice.py:
```diff
@@ -20,40 +21,39 @@
-from tools.translate import _
 
+from openerp.models import Model, api, _
+from openerp.fields import Many2one
```
Instead this limited scope, we use to import `from openerp import fields`, and then declare fields with `fields.Many2one` so that if you need to add another field type in the future, don't need to change the import

pedrobaeza commented on crm_claim_rma/account_invoice.py:
```diff
@@ -20,40 +21,39 @@
         """ Override when from claim to update the quantity and link to the
         claim line."""
+
+        context = self.env.context
```
This is not needed anymore with new API

pedrobaeza commented on crm_claim_rma/account_invoice.py:
```diff
@@ -72,38 +72,41 @@ def _refund_cleanup_lines(self, cr, uid, lines, context=None):
+    def _prepare_refund(self, invoice, date=None, period_id=None,
+                        description=None, journal_id=None):
+
+        context = self.env.context
```
The same

pedrobaeza commented on crm_claim_rma/crm_claim_rma.py:
```diff
@@ -0,0 +1,584 @@
+##############################################################################
+
+from openerp.models import Model, api, _
+from openerp.fields import (Char, Date, Float, One2many, Many2one, Selection,
```
See my previous comment for this

pedrobaeza commented on crm_claim_rma/crm_claim_rma.py:
```diff
@@ -0,0 +1,584 @@
+
+    def init(self, cr):
+        cr.execute("""
+            UPDATE "crm_claim" SET "number"=id::varchar
```
This should be done in a pre-init hook or better, depend on the new crm_claim_code module: https://github.com/OCA/crm/tree/8.0/crm_claim_code

aristobulo-eezee replied:
@pedrobaeza sorry, first time I hear the concept of pre-init hooks in Odoo, could you tell me where to find more info about it?

pedrobaeza replied:
It's the way to make one task in module installation. It's done in the module I mention that I recommend you to use instead of rewriting this: https://github.com/OCA/crm/blob/8.0/crm_claim_code/__init__.py#L10 and declare here: https://github.com/OCA/crm/blob/8.0/crm_claim_code/__openerp__.py#L48

aristobulo-eezee committed:
Use wide scope when importing fields.

nhomar commented:
@aristobulo-eezee 

I insist.

Why are you redoing a job that is done already here:

https://github.com/OCA/rma/pull/26

You can test our fork (which is the one proposed for merge) on https://github.com/Vauxoo/rma

Only 1 pylint error until now (check the travis link in the PR).

The PR have teh description and it is a WiP yet with a huge time investment >> 400 hours of work.

I do not understand why you are redoing this job.

For me is a :-1: (if somebody else is not in faver of our PR please put a :-1: there.

Regards.

aristobulo-eezee commented:
@nhomar before starting this port I asked if there was already any other ongoing plan about this repo, as I got no response we started our work (we didn't knew the existence of your branch). 

Our work was based on original crm_claim_rma and was only to make it run on v8 as we need, not adding new features (because in our case, we don't need it), and as we just made it run again in v8 we wanted to contribute what we did to OCA, that's it.

If the OCA wants to merge yours, then ok.

nhomar commented:
2015-07-31 4:56 GMT-05:00 Aristobulo Meneses notifications@github.com:

> @nhomar https://github.com/nhomar before starting this port I asked if
> there was already any other ongoing plan about this repo, as I got no
> response we started our work (we didn't knew the existence of your branch).

@ both miss the communication sorry.....

Please test our branch (this module was almost untouched also, but fixed
some conceptual problems).

The feedback is welcome and I know you will enojoy such job.

Thanks for the contribution BTW.

Regards.

## 

---

Saludos Cordiales

CEO at Vauxoo https://www.vauxoo.com Odoo's Gold Partner.

[image: --]
Nhomar Hernandez
[image: http://]about.me/nhomar
http://about.me/nhomar?promo=email_sig

aristobulo-eezee committed:
Import fields using a wide scope. Update fields declaration

aristobulo-eezee committed:
Merge branch 'port_crm_claim_rma' of https://github.com/eezee-it/rma into port_crm_claim_rma

aristobulo-eezee committed:
Proper use of context with new api

aristobulo-eezee committed:
Get context directly from self.env

aristobulo-eezee committed:
Fix fields declaration

aristobulo-eezee committed:
Use pre_init_hook to fill claim number

aristobulo-eezee committed:
Fix line under-indented issue

aristobulo-eezee committed:
Add missing blank lines in __init__.py

aristobulo-eezee committed:
Fix pre_init_hook parameter

aristobulo-eezee committed:
Modify pre_init_hook method

aristobulo-eezee committed:
Use crm_claim_code

aristobulo-eezee committed:
Remove unused ValidationError import. Remove unnecesary blank lines

aristobulo-eezee committed:
Add OCA/crm dependency

aristobulo-eezee commented:
@pedrobaeza I have added OCA/crm in travis.yml to but is not available in runbot. How can this be fixed?

pedrobaeza commented:
I have added the dependency and restarted the build.

aristobulo-eezee commented:
Thank you @pedrobaeza

nhomar commented on crm_claim_rma/wizard/claim_make_picking.py:
```diff
@@ -0,0 +1,279 @@
+
+        return good_lines
+
+    claim_line_source_location = Many2one(
```
fields.Many2one

nhomar commented on crm_claim_rma/res_partner_view.xml:
```diff
@@ -15,7 +15,7 @@
           <field name="is_company" invisible="1"/>
-          <field name="country" invisible="1"/>
+          <!--<field name="country" invisible="1"/>-->
           <field name="country_id"/>
```
Delete the comment.

yaniaular commented on crm_claim_rma/crm_claim_rma.py:
```diff
@@ -0,0 +1,550 @@
+            res.append((claim.id, '[' + code + '] ' + claim.name))
+        return res
+
+    def copy_data(self, cr, uid, id, default=None, context=None):
```
@aristobulo-eezee The parameters, Can you place as you did in line <a href="https://github.com/OCA/rma/pull/32/files#diff-cfd8c4bc0a8c5905769c8550af1271b4R78">78</a> ? and add @api.model. Thanks

yaniaular commented on crm_claim_rma/wizard/claim_make_picking.py:
```diff
@@ -0,0 +1,279 @@
+    _description = 'Wizard to create pickings from claim lines'
+
+    # Get default source location
+    def _get_source_loc(self):
```
@aristobulo-eezee I think that this method should have the @ api.model because it is called from a field of API 8.0  and without @api.model the default not working. Did you do a functional test in a test instance about this method for default of claim_line_source_location field?

yaniaular commented on crm_claim_rma/wizard/claim_make_picking.py:
```diff
@@ -0,0 +1,279 @@
+        return loc_id
+
+    # Get default destination location
+    def _get_dest_loc(self):
```
@aristobulo-eezee I think that this method should have the @api.model because it is called from a field of API 8.0 and the method without @api.model not working. Did you do a functional test in a test instance about this method for default of  claim_line_dest_location field?

After the method have @api.model, the method must return records. So, loc_id = False, must be loc_id = self.env['stock.location'] .

nhomar commented:
@aristobulo-eezee we will use your approach here, can you tell us if you will finish the comments done by @yaniaular in order to move forward all the squash on the other huge PR?

regards and really thanks for the contribution.

aristobulo-eezee commented:
@nhomar yes, I plan to do it this weekend.

LeartS commented:
@aristobulo-eezee I can work on it today & tomorrow, can I make PRs against your branch?

nhomar commented:
@aristobulo-eezee Can you merge the changes that @LeartS will do in the week?

We are against the clock here due to your approach is one of the main ones :-)

PS: @LeartS we can do it also we are making the adaptation to our modules to this one already in the big PR. then just let me know to trigger the work here.

Regards.

aristobulo-eezee commented:
@LeartS of course ;)

LeartS committed:
Fix imports, docstrings and other conventions

Imports https://www.python.org/dev/peps/pep-0008/#id17
and http://odoo-new-api-guide-line.readthedocs.org/en/latest/conventions.html

Docstrings https://www.python.org/dev/peps/pep-0257/

LeartS committed:
Follow OCA guidelines for module structure

https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md

mistotebe commented:
Can you pick a couple of changes we made in #24? I've taken a few snapshots from yours and annotated the UI changes we made:
![claim_form](https://cloud.githubusercontent.com/assets/1913139/9084115/03897e1a-3b68-11e5-8150-411eb7a4d567.png)
![claim_line](https://cloud.githubusercontent.com/assets/1913139/9084116/03b044e6-3b68-11e5-8c77-e007d8662ed3.png)

Other changes include: UoM support https://github.com/credativUK/rma/commit/90890a6abc15b16ee3e23799e616b7619f72a482, claim (line) value https://github.com/credativUK/rma/commit/beff4c02b7d523e725aa1202cd7489aa7e8dbc37 and https://github.com/credativUK/rma/commit/dbbe36ad051737a0a1d78e3f733dd1e94f029af6, proper unittests https://github.com/credativUK/rma/commit/b55258dd7e53e12849cf555827ea6d35468d3897, typos fixed in https://github.com/credativUK/rma/commit/52f9ad4d12e1763835a2887abcf0bccf26b405a8, possibly others.

If you want, I could make them and propose for you to merge, probably not today, however. I've retriggered the build on runbot which got expired since so that you can see what #24 looks like live.

LeartS commented:
@mistotebe I'm working on this PR right now, I'll try to integrate some of your changes

nhomar commented:
@mistotebe absolutly but they looks familiar to me I will check with @yaniaular  in a couple of hours.

mistotebe commented:
Sure, these are changes that I remember our client needed so we can't actually switch to this module without those as that would be considered a regression by them. Most should be quite straightforward, some might even be cherry-pickable. Feel free to pick stuff from #24 as needed.

The rebuild of #24 has finished and it's available on runbot again.

nhomar commented:
@LeartS @mistotebe 

On this PR https://github.com/OCA/rma/pull/34/files which include a super wizard to load rma line magically the return lines parts stay obsollete, we are working on this PR to make it perfect.

Basically in our use case the customer asked for a multi-search windows that can be triggered by a barcode scanner and look by product, serial and/or customer... this specific part can be done separatelly.

The concept of unit is not necesary there due to the unit of measure is taken from the invoice line itself.

@LeartS  The other point about the positioning of buttons is good and clear.! (I think we didi in our version of the module here in vauxoo/rma#8.0 you can pick the xml there.

LeartS commented:
@nhomar so should I continue working on this PR or should I work on yours? I'm slightly confused right now.

nhomar commented:
On this one (commit and rebase whatever here).

But you can pick from Vauxoo the xml if you wish and from @mistotebe ' s ones what you need but continue here which is the one eith everything migrated to new api (@mistotebe and mine are not).

Regards and thanks.

LeartS commented:
Ok, perfect. Thanks

mistotebe commented:
@nhomar, how does the following scenario work with the improvements you mention:
- a product is sent over to customer and has quite a long warranty
- there is a problem with the product and spare parts (not a new product) have to be sent over
- the value and types/quantities of items sent over should be reportable per customer and product serial number

Is that still possible? At the moment, this is only doable with creating new lines

nhomar commented:
Hello @mistotebe 

It is in our roadmap and it is fairly done (a WIP).

On our scenario we used the mrp_repair (official) module and link such module to the claim.

The mrp_repair module manage that concept of include costs and invoice parts to customer or simply give them for free.

Then basically:
1. You receive/declare the product exactly as you sell it.
2. You move the task of review and reporting to repair automagically.
3. Repair make its job (declare the need to send parts and so on).
4. Repair send to rma (which is in contact with the customer).
5. End.

I hope it clarify?

PS: about the serial number part it is a WiP becuase we are designing the solution, to use serial parts we need to be sure to sell products with components which were serialized in the out itself (and that is not supported by odoo Out of the box right now).`<< If you know some pùblic apporach for that it is welcome>>`

mistotebe commented:
If it's not ready yet, it might be better to just copy the code that we wrote since the rest can still be plugged in fine.

As far as mrp_repair goes, that seems too complex for this scenario, where this starts as a claim anyway and there are just some parts sent over if there is no repair done by an engineer and the claim.line/crm.claim already store all the required information for subsequent reporting as well.

LeartS committed:
Improve port to new API and fix exceptions error.

Simplify some methods, work with recordset instead of ids when
possible, use @api.one, follow naming guidelines, ...

Also fixes an error introduced in the previous-to-last commit
regarding the custom exceptions.

LeartS commented:
I'm working on this on my branch https://github.com/LeartS/rma/tree/port_crm_claim_rma_fixes
I created a PR (still WIP but every commit should be deployable and reviewable) against the branch of this PR here: https://github.com/eezee-it/rma/pull/4 comments are welcome there.

mistotebe committed:
Move claim buttons to the header

LeartS committed:
Improve crm.claim views

Put buttons in the header, fix layout (smart buttons),
reduce code duplication, remove duplicated name field,
fix indentation and whitespace,
rename 'Return lines' to 'claim lines', ...

LeartS committed:
Ensure recordset is a single record in methods, slight adjustments

LeartS committed:
Fix some outdated references to the renamed fields

LeartS committed:
Show location labels in return wizard, remove commented code

LeartS committed:
Work with recordsets instead of ids

In the `_get_common_*` methods and also in the default functions

LeartS commented:
My PR to this branch is ready for review, see comment: https://github.com/eezee-it/rma/pull/4#issuecomment-128695112

EDIT: @aristobulo-eezee has merged my changes, so we can review here

aristobulo-eezee committed:
Merge pull request #4 from LeartS/port_crm_claim_rma_fixes

Improvements to the port of crm_claim_rma

nhomar commented:
@osvalr @yaniaular can you make a proper review here please testing mixed with vauxoo/rma ports please

LeartS committed:
Fix PEP8 warnings

LeartS committed:
Fix error on installation related to order of xml records

The claim form view refers to some window actions that were declared
after. Move the view definition after the actions definition to
fix error.

yaniaular commented:
@aristobulo-eezee @pedrobaeza @nhomar 

It sounds good the changes in this PR. :+1: <b>BUT</b> it should be merged the corrections of @LeartS in https://github.com/eezee-it/rma/pull/5 before this PR https://github.com/OCA/rma/pull/32

We will be taking the enhancements from here https://github.com/learts/rma/tree/crm_claim_rma-fix-travis-errors into https://github.com/vauxoo-dev/rma/tree/8.0-squash_crm_claim_rma and we will create a new PR to the OCA later.

@osvalr What do you think?

osvalr commented:
@yaniaular :+1: I agree

aristobulo-eezee committed:
Merge pull request #5 from LeartS/crm_claim_rma-fix-travis-errors

Fix travis errors

nhomar commented:
Travis green 2 +1 merging to continue the migration #26
